workspace.go

```diff
@@ -0,0 +1,390 @@
+		// Determine repo root.
+		// This is potentially somewhat slow.
+		bpkg, err := build.Import(importPath, "", build.FindOnly)
+		if err != nil {
+			log.Println("build.Import:", err)
+			continue
+		}
+		if bpkg.Goroot {
+			// Go-Package-Store has no support for updating packages in GOROOT, so skip those.
+			continue
+		}
+		vcsCmd, root, err := vcs.FromDir(bpkg.Dir, bpkg.SrcRoot)
+		if err != nil {
+			// Go package not under VCS.
+			continue
+		}
+		vcs, err := vcsstate.NewVCS(vcsCmd)
+		if err != nil {
+			log.Printf("repo %v not supported by vcsstate: %v", root, err)
+			continue
+		}
+
+		var repo *pkg.Repo
+		w.reposMu.Lock()
+		if _, ok := w.repos[root]; !ok {
+			repo = &pkg.Repo{
+				Path: bpkg.Dir,
+				Root: root,
+				Cmd:  vcsCmd,
+				VCS:  vcs,
+				// TODO: Maybe keep track of import paths inside, etc.
+			}
+			w.repos[root] = repo
+		} else {
+			// TODO: Maybe keep track of import paths inside, etc.
+		}
+		w.reposMu.Unlock()
+
+		// If new repo, send off to phase 2 channel.
+		if repo != nil {
+			w.unique <- repo
+		}
+	}
+}
+
+// importPathRevisionWorker sends unique repositories to phase 2.
+func (w *workspace) importPathRevisionWorker(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for p := range w.importPathRevisions {
+		//started := time.Now()
+		// Determine repo root.
+		// This is potentially somewhat slow.
+		rr, err := vcs.RepoRootForImportPath(p.importPath, false)
+		if err != nil {
+			panic(err) // TODO.
+		}
+		//fmt.Printf("rr: %v ms.\n", time.Since(started).Seconds()*1000)
+		remoteVCS, err := vcsstate.NewRemoteVCS(rr.VCS)
+		if err != nil {
+			log.Printf("repo %v not supported by vcsstate: %v\n", rr.Root, err)
+			continue
+		}
+
+		var repo *pkg.Repo
+		w.reposMu.Lock()
+		if _, ok := w.repos[rr.Root]; !ok {
+			repo = &pkg.Repo{
+				Root:      rr.Root,
+				RemoteURL: rr.Repo,
+				Cmd:       rr.VCS,
+				RemoteVCS: remoteVCS,
+				// TODO: Maybe keep track of import paths inside, etc.
+			}
+			repo.Local.Revision = p.revision
+			w.repos[rr.Root] = repo
+		} else {
+			// TODO: Maybe keep track of import paths inside, etc.
+		}
+		w.reposMu.Unlock()
+
+		// If new repo, send off to phase 2 channel.
+		if repo != nil {
+			w.unique <- repo
+		}
+	}
+}
+
+// processFilterWorker computes repository remote revision (and local if needed)
+// in order to figure out if repositories should be presented.
+func (w *workspace) processFilterWorker(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for p := range w.unique {
+		//started := time.Now()
+		// Determine remote revision.
+		// This is slow because it requires a network operation.
+		var remoteRevision string
+		if p.VCS != nil {
+			var err error
+			remoteRevision, err = p.VCS.RemoteRevision(p.Path)
+			_ = err // TODO.
+		} else if p.RemoteVCS != nil {
+			var err error
+			remoteRevision, err = p.RemoteVCS.RemoteRevision(p.RemoteURL)
+			_ = err // TODO.
+		}
+		//fmt.Printf("remoteVCS.GetRemoteRev: %v ms.\n", time.Since(started).Seconds()*1000)
```

Comment:
Presuming you have this here to benchmark this in the future.. one day we'll be able to replace things like this with simple and easy appdash tracing :)